src/box/lua/net_box.c

```diff
@@ -2382,8 +2446,12 @@ static uint32_t
 netbox_transport_fetch_schema(struct netbox_transport *transport,
 			      struct lua_State *L)
 {
-	transport->state = NETBOX_FETCH_SCHEMA;
-	netbox_transport_on_state_change(transport, L);
+	if (transport->state != NETBOX_GRACEFUL_SHUTDOWN) {
+		assert(transport->state == NETBOX_AUTH ||
+		       transport->state == NETBOX_ACTIVE);
+		transport->state = NETBOX_FETCH_SCHEMA;
+		netbox_transport_on_state_change(transport, L);
+	}
```

Comment:
Just didn't want to write a test, hoped you wouldn't notice. Fixed.

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/net_box.c b/src/box/lua/net_box.c
index 9a893d7ed103..67f6eb138f88 100644
--- a/src/box/lua/net_box.c
+++ b/src/box/lua/net_box.c
@@ -2444,14 +2444,19 @@ netbox_transport_do_auth(struct netbox_transport *transport,
  */
 static uint32_t
 netbox_transport_fetch_schema(struct netbox_transport *transport,
-			      struct lua_State *L)
+			      struct lua_State *L, uint32_t schema_version)
 {
-	if (transport->state != NETBOX_GRACEFUL_SHUTDOWN) {
-		assert(transport->state == NETBOX_AUTH ||
-		       transport->state == NETBOX_ACTIVE);
-		transport->state = NETBOX_FETCH_SCHEMA;
-		netbox_transport_on_state_change(transport, L);
+	if (transport->state == NETBOX_GRACEFUL_SHUTDOWN) {
+		/*
+		 * If a connection is in the 'graceful_shutdown', it can't
+		 * issue new requests so there's no need to fetch schema.
+		 */
+		return schema_version;
 	}
+	assert(transport->state == NETBOX_AUTH ||
+	       transport->state == NETBOX_ACTIVE);
+	transport->state = NETBOX_FETCH_SCHEMA;
+	netbox_transport_on_state_change(transport, L);
 	uint32_t peer_version_id = transport->greeting.version_id;
 	/* _vcollation view was added in 2.2.0-389-g3e3ef182f */
 	bool peer_has_vcollation = peer_version_id >= version_id(2, 2, 1);
@@ -2471,7 +2476,7 @@ restart:
 	bool got_vspace = false;
 	bool got_vindex = false;
 	bool got_vcollation = false;
-	uint32_t schema_version = 0;
+	schema_version = 0;
 	do {
 		struct xrow_header hdr;
 		if (netbox_transport_send_and_recv(transport, &hdr) != 0)
@@ -2541,9 +2546,10 @@ restart:
 
 /**
  * Processes iproto requests in a loop until an error or a schema change.
- * Returns on a schema change. On failure raises a Lua error.
+ * Returns the current schema version on schema change. On failure raises
+ * a Lua error.
  */
-static void
+static uint32_t
 netbox_transport_process_requests(struct netbox_transport *transport,
 				  struct lua_State *L, uint32_t schema_version)
 {
@@ -2559,7 +2565,7 @@ netbox_transport_process_requests(struct netbox_transport *transport,
 		netbox_transport_dispatch_response(transport, L, &hdr);
 		if (hdr.schema_version > 0 &&
 		    hdr.schema_version != schema_version) {
-			break;
+			return hdr.schema_version;
 		}
 	}
 }
@@ -2573,10 +2579,12 @@ netbox_connection_handler_f(struct lua_State *L)
 	struct netbox_transport *transport = (void *)lua_topointer(L, 1);
 	netbox_transport_do_id(transport, L);
 	netbox_transport_do_auth(transport, L);
+	uint32_t schema_version = 0;
 	while (true) {
-		uint32_t schema_version;
-		schema_version = netbox_transport_fetch_schema(transport, L);
-		netbox_transport_process_requests(transport, L, schema_version);
+		schema_version = netbox_transport_fetch_schema(
+			transport, L, schema_version);
+		schema_version = netbox_transport_process_requests(
+			transport, L, schema_version);
 	}
 	return 0;
 }
diff --git a/test/box-luatest/graceful_shutdown_test.lua b/test/box-luatest/graceful_shutdown_test.lua
index da510f7f2ceb..87de083d45f3 100644
--- a/test/box-luatest/graceful_shutdown_test.lua
+++ b/test/box-luatest/graceful_shutdown_test.lua
@@ -338,3 +338,91 @@ g.test_shutdown_aborted_on_reconnect_2 = function()
     t.assert_equals(conn.state, 'error_reconnect')
     conn:close()
 end
+
+g.after_test('test_schema_not_fetched_during_shutdown', function()
+    if not g.server.process then
+        g.server:start()
+    end
+    g.server:exec(function()
+        if box.space.test then
+            box.space.test:drop()
+        end
+    end)
+end)
+
+-- Checks that a connection that entered the 'graceful_shutdown' state stops
+-- fetching schema.
+g.test_schema_not_fetched_during_shutdown = function()
+    g.server:exec(function()
+        local fiber = require('fiber')
+        rawset(_G, 'cond1', fiber.cond())
+        rawset(_G, 'wait1', function()
+            box.session.push(true)
+            _G.cond1:wait()
+            return true
+        end)
+        rawset(_G, 'signal1', function()
+            _G.cond1:broadcast()
+        end)
+        rawset(_G, 'cond2', fiber.cond())
+        rawset(_G, 'wait2', function()
+            box.session.push(true)
+            _G.cond2:wait()
+            return true
+        end)
+        rawset(_G, 'signal2', function()
+            _G.cond2:broadcast()
+        end)
+        box.ctl.set_on_shutdown_timeout(9000)
+    end)
+    local conn1 = net.connect(g.server.net_box_uri)
+    local on_trigger_start = fiber.channel(1)
+    local stop_trigger = fiber.channel(1)
+    conn1:on_shutdown(function()
+        on_trigger_start:put(true)
+        stop_trigger:get()
+    end)
+    local conn2 = net.connect(g.server.net_box_uri)
+    local schema_reloaded = false
+    conn2:on_schema_reload(function()
+        schema_reloaded = true
+    end)
+    local fut1 = conn2:call('wait1', {}, {is_async = true})
+    -- Wait for the server to invoke the call.
+    for _, v in fut1:pairs() do -- luacheck: ignore
+        t.assert_equals(v, true)
+        break
+    end
+    local fut2 = conn2:call('wait2', {}, {is_async = true})
+    -- Wait for the server to invoke the call.
+    for _, v in fut2:pairs() do -- luacheck: ignore
+        t.assert_equals(v, true)
+        break
+    end
+    local on_server_stop = fiber.channel(1)
+    fiber.create(function()
+        g.server:stop()
+        on_server_stop:put(true)
+    end)
+    t.assert(on_trigger_start:get())
+    t.assert(conn1.state, 'active')
+    t.assert(conn2:wait_state('graceful_shutdown'))
+    conn1:eval([[box.schema.space.create('test')]])
+    conn1:call('signal1')
+    fiber.sleep(0.01)
+    t.assert(conn1.space.test)
+    t.assert_not(conn2.space.test)
+    t.assert_not(schema_reloaded)
+    conn1:eval([[box.space.test:drop()]])
+    conn1:call('signal2')
+    fiber.sleep(0.01)
+    t.assert_not(conn1.space.test)
+    t.assert_not(conn2.space.test)
+    t.assert_not(schema_reloaded)
+    stop_trigger:put(true)
+    t.assert(on_server_stop:get())
+    t.assert(fut1:result())
+    t.assert(fut2:result())
+    conn1:close()
+    conn2:close()
+end
```

</details>